### PR TITLE
Fix false positives in unused private members

### LINKED ISSUE #812: UnusedPrivateMember false positives
version: 1.0.0.RC6-4
**Case 1**

```kotlin
@ConsumerScope
@Component(dependencies = arrayOf(OneComponent::class))
interface AnotherComponent {
    fun inject(myActivity: MyActivity)
}
```
Rule complains about `myActivity`.

```kotlin
interface UserPlugin {
    fun plug(application: Application)
    fun unplug()
}
```

Rule complains about `application`

**Case 2**

```kotlin
    override fun funA() {
        objectA.resolve(valA, object : MyCallback {
            override fun onResolveFailed(throwable: Throwable) {
                errorMessage.visibility = View.VISIBLE
            }
        })
    }
```

rule complains about `throwable`

**Case 3**
```kotlin 
class A {
    companion object {
        private const val MY_CONST = 42
    }

    fun a() {
        Compleatable.timer(MY_CONST.toLong(), TimeUnit.MILLISECONDS)
              .subscribe()
    }
}
```

rule complains about `MY_CONST`

PS. I have many more violations in my codebase. I just randomly picked few of them, but most likely all of them a similar.

### PR DESCRIPTION
**Pull request by arturbosch**
-  This should fix all false positives in  #812 
Edit: we can't turn on the rule in our code base for now due to detektCheck

### CODE REVIEW

**Comment by Mygod** [comment]
This still doesn't fix false positives for extension private functions. For example this:

https://github.com/Mygod/VPNHotspot/blob/ed4c9a5610d447ee4e7541f45ca85fe3674661d9/mobile/src/main/java/be/mygod/vpnhotspot/RepeaterService.kt#L52-L55

**Comment by Mygod** [comment]
Also private members in companion objects: https://github.com/shadowsocks/shadowsocks-android/blob/a7d7746195d687619899ac7b73fed7791eea9492/mobile/src/main/java/com/github/shadowsocks/AppManager.kt#L65

**Comment by Mygod** [comment]
Also variables named `_`: https://github.com/shadowsocks/shadowsocks-android/blob/a7d7746195d687619899ac7b73fed7791eea9492/mobile/src/main/java/com/github/shadowsocks/ProfilesFragment.kt#L278

**Comment by arturbosch** [comment]
Hey, thanks for the feedback. I will look into your cases.

**Comment by vanniktech** [comment]
We also probably want to exclude properties that are annotated with @Rule by default since sometimes they do some work but you don't interact on them.

**Comment by schalkms** [comment]
@vanniktech For now, this would make sense. It is in the VCS anyway.

**Comment by arturbosch** [comment]
@schalkms this kind of rule just needs symbol solving when also considering public members, it is alright to have such a rule now. The problem was, that you can declare many additional objects in the same file and use their constants which was not considered in the initial implementation.
It's not that I need so long with this PR it's just that I do not have time :(

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/Case.kt [hunk 1]
```diff
@@ -77,14 +77,14 @@ enum class Case(val file: String) {
 	TrailingWhitespaceNegative("/cases/TrailingWhitespaceNegative.kt"),
 	TrailingWhitespacePositive("/cases/TrailingWhitespacePositive.kt"),
 	NoTabsNegative("/cases/NoTabsNegative.kt"),
-	NoTabsPositive("/cases/NoTabsPositive.kt");
+	NoTabsPositive("/cases/NoTabsPositive.kt"),
+	UnusedPrivateMembers("/cases/UnusedPrivateMembers.kt");
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/Case.kt]: Should we stick to splitting **positive** and **negative** test cases also for `UnusedPrivateMembers`?

Comment (arturbosch) [reply detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/Case.kt]: I will try this later.

**Comment by arturbosch** [comment]
Besides detekt, spek and arrow, I need like 2-3 more projects for QA.
Feel free to name some.

**Comment by schalkms** [comment]
@arturbosch the detekt compiler project and ktlint

**Comment by schalkms** [comment]
@vanniktech definitely good idea! +1

**Comment by arturbosch** [comment]
Ok, also added a configurable to allow unused private members matching  a regex (perconfigured with _|ignored|expected).

Jap, a list with projects using detekt would be cool to test out new rules/features and make sure nothing breaks ^^

**Pull request merged**